### PR TITLE
AST: Add scaffolding and definitions for the AST definition macro

### PR DESCRIPTION
**Pull request by kontheocharis**
The macro is named `define_tree!` and it essentially takes all the items of the AST definition and (will eventually) produce visitor and walker definitions for them. This is an intermediate PR to keep the code size small; the parsing has been implemented but the emitting has not.

Working towards #519.

### CODE REVIEW

File: compiler/hash-tree-def/src/parsing.rs [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+/// Helper type so that we can implement [`TryFrom<&Item>`] for optional
+/// [`TreeDefOpts`].
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct MaybeTreeDefOpts(Option<TreeDefOpts>);
+
+impl TryFrom<&Item> for MaybeTreeDefOpts {
+    type Error = syn::Error;
+    fn try_from(value: &Item) -> Result<Self, Self::Error> {
+        // Options are given by a macro define_tree_opts! { ... }
+        println!("{:#?}", value);
```
Comment (feds01) [inline compiler/hash-tree-def/src/parsing.rs:224]: remove?

File: compiler/hash-tree-def/src/parsing.rs [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+                        let node_name = ensure_single_type_argument(segment)?;
+                        Ok(StructNodeFieldData::ChildList { node_name })
+                    } else {
+                        Ok(StructNodeFieldData::Other { ty: value.clone() })
+                    }
+                }
+                _ => Ok(StructNodeFieldData::Other { ty: value.clone() }),
+            },
+            _ => Ok(StructNodeFieldData::Other { ty: value.clone() }),
+        }
```
Comment (feds01) [inline compiler/hash-tree-def/src/parsing.rs:129]: would be nice to use `if_chain!` here but not necessarily required.

**Review by feds01 — CHANGES_REQUESTED** [review]
mui bien sie

**Review by feds01 — APPROVED** [review]

File: compiler/hash-tree-def/src/parsing.rs (hunk 1 above)
Comment (kontheocharis) [reply compiler/hash-tree-def/src/parsing.rs:224]: Will tackle in final PR

**Pull request merged**